hgshah/cmput404-project
# Check if remote author follows local author

## Changes
- New endpoint: https://turnip-oomfie-1.herokuapp.com/api/schema/redoc/#tag/follows/operation/authors_followers_retrieve_2
- Fix AuthorSerializer
- Add some debugging stuff in local.py and local_mirror.py
- Fix the methods in FollowersView: port from Author.objects.get to Author.get_author
- Fix a lot of tests
- Create auto tester for the follow flow
- Create a 'author/self' endpoint to figure out who's currently logged in

## Automated testing (easier for testing)
1. Delete all Follow objects related to actor in 8000 and target in 8080. Delete the Follow objects in /admin for each port.
2. Make sure that our server is running in 8000 and 8080
3. run `python manage.py test remote_nodes/test_mirror`. The script does the stuff below: ([link](https://github.com/hgshah/cmput404-project/pull/99/files#diff-1e28a49e9feaa63b0a3630e08a90aa6782b059470cd33e370759807d715cf1aaR1))
    - check if 8000 actor follows 8080 target
    - 8000 actor follows 8080 target
    - 8080 target sees 8000's request
    - check if 8000 actor follows 8080 target

## TODO:
- [x] Make sure inbox, comments, post still works
- [x] Fix all tests
- [x] Make sure recursive author still works

## Manual testing (optional)
1. run two local servers 8000 and 8080
4. get an author url from http://127.0.0.1/8080/authors/ -> let's call this remote
2. get an author url from http://127.0.0.1/8000/authors/ -> let's call this local
5. in http://127.0.0.1/8000/admin/, make a new follow object with the following fields
    - actor: the url from http://127.0.0.1/8080/ (remote)
    - target: the url from http://127.0.0.1/8000/ (local)
    - has_accepted: True
6. check if remote follows local at http://127.0.0.1:8000/authors/<local-id>/followers
7. check if remote follows local at http://127.0.0.1:8000/authors/<local-id>/followers/<remote-id>/

## Full testing flow (context; not yet complete)
1. run two local servers 8000 and 8080
2. get an author url from http://127.0.0.1/8080/authors/ -> let's call this remote
2. get an author url from http://127.0.0.1/8000/authors/ -> let's call this local
4. in 8000, login as local, and go to http://127.0.0.1:8000/authors/<remote-id>/followers/
5. Do POST
6. in 8080, login as remote, go to http://127.0.0.1:8080/follows/incoming/
8. click the localUrl in the follow request made by local
9. Check hasAccepted, and do PUT
10. local should now be following remote
11. remote can confirm this by looking at http://127.0.0.1:8080/authors/<remote-id>/followers
12. local can confirm this by looking at http://127.0.0.1:8000/authors/<remote-id>/followers

## Screenshot

![image](https://user-images.githubusercontent.com/21351900/202838246-9bdbeaaf-0171-4bec-8cb9-795d6b24bdc9.png)


BELOW IS NO LONGER DOABLE
> Check out super from turnip's followers at: http://turnip-oomfie-1.herokuapp.com/authors/dccab2ec-72e3-4a0e-9aff-83b1b851e794/followers/
>
> Let's see if super from potato follows super from turnip at: http://turnip-oomfie-1.herokuapp.com/authors/dccab2ec-72e3-4a0e-9aff-83b1b851e794/followers/55894dfd-6f71-4215-8400-27b57d1d920d
> - Log in as potato pretending to be the potato server. Username and password is potato.
